vpr/src/analytical_place/full_legalizer.cpp

```diff
@@ -281,6 +291,710 @@ static LegalizationClusterId create_new_cluster(PackMoleculeId seed_molecule_id,
+        // full and try to cluster the unclustered molecules in same tile again.
+        if (!illegal_cluster_mols.empty()) {
+            cluster_legalizer.set_legalization_strategy(ClusterLegalizationStrategy::FULL);
+            created_clusters.clear();
+            cluster_molecules_in_tile(tile_loc,
+                                    tile_type,
+                                    illegal_cluster_mols,
+                                    cluster_legalizer,
+                                    primitive_candidate_block_types,
+                                    created_clusters);
+
+            // Clean clusters created with full strategy not to increase memory footprint.
+            for (LegalizationClusterId cluster_id : created_clusters) {
+                cluster_legalizer.clean_cluster(cluster_id);
+            }
+        }
+    }
+}
+
+void FlatRecon::neighbor_clustering(ClusterLegalizer& cluster_legalizer,
+                                    const vtr::vector<LogicalModelId, std::vector<t_logical_block_type_ptr>>& primitive_candidate_block_types,
+                                    std::unordered_set<PackMoleculeId>& mols_clustered) {
+    vtr::ScopedStartFinishTimer neigh_pass_clustering("Neighbor Pass Clustering");
+    
+    // Iterate over molecules and try to join the unclustered ones to their
+    // already created 8-neighboring tile clusters.
+    for (APBlockId blk_id : ap_netlist_.blocks()) {
+        // Get unclustered block and its location.
+        PackMoleculeId molecule_id = ap_netlist_.block_molecule(blk_id);
+        t_physical_tile_loc loc = mol_desired_physical_tile_loc[molecule_id];
+
+        // Skip the already clustered molecules.
+        if (cluster_legalizer.is_mol_clustered(molecule_id))
+            continue;
+
+        // Skip the io blocks in that pass. This aims to avoid unneccessary recreation of clusters neighboring io blocks.
+        // TODO: This might be handled more generally instead of checking block name.
+        t_logical_block_type_ptr block_type = infer_molecule_logical_block_type(molecule_id, prepacker_, primitive_candidate_block_types);
+        std::string block_name = block_type->name;
+        if (block_name == "io")
+            continue;
+
+        // Get 8-neighbouring tile locations of the current molecule in the same layer.
+        std::vector<t_physical_tile_loc> neighbor_tile_locs;
+        neighbor_tile_locs.reserve(8);
+        auto [layers, width, height] = device_grid_.dim_sizes();
+        for (int dx : {-1, 0, 1}) {
+            for (int dy : {-1, 0, 1}) {
+                if (dx == 0 && dy == 0) continue;
+                int neighbor_x = loc.x + dx, neighbor_y = loc.y + dy;
+                if (0 <= neighbor_x && neighbor_x < (int)width  && 0 <= neighbor_y && neighbor_y < (int)height) {
+                    neighbor_tile_locs.push_back({neighbor_x, neighbor_y, loc.layer_num});
+                }
```

Comment:
If you're just trying to skip incompatible tiles, you could do it here instead of with the hacky IO code above.  But it may not matter much, as you check that in the callee.